### ISSUE TITLE
[performance] `moveToActive` should only be called once per queue a job notification

### ISSUE THREAD

**Issue by bradvogel** [OptimalBits/bull#557]
Moving conversation from https://github.com/OptimalBits/bull/issues/553#issuecomment-304788478 here

We currently have an inefficiency where when a queue receives a new job notification, it calls the `moveToActive` script `concurrency` times. This creates a lot of extra Redis traffic (and Node CPU burn) for queues with high-concurrency.

To repro:
1) Above line https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L719, add:
```
          .then(job => {
            console.log('getNextJob',!!job)
            return job;
          })
```
2) Instantiate a new queue with concurrency=40
3) Submit a job to that queue

You'll see `getNextJob true` logged 1 time and `getNextJob false` get logged 39 times.

A better approach is to have the queue itself listen for job notifications and then dispatch it to an available worker.

**Comment by manast** [comment]
Yeah, it is a bit tricky actually. I was thinking about an approach where you could have a special event emitter that only emits to one of its listeners. For that to work I would need to setup the listener directly after the call to moveToActive when it returns null instead of a job(indicating that no job was in the queue, however I am not sure if a job could slip between the returning of moveToActive and my listener setup. If this does not happen vey often, the job that slips would be catched by the guard a few seconds later... I have a vague memory that some tests failed with a timeout if I moved the listener setup to after moveToActive, I will make some experiments and let you know.

**Comment by bradvogel** [comment]
You're right. This will take some rearchitecting of how jobs are processed. I can take a look tomorrow.

**Comment by bradvogel** [comment]
@ randallm working on this

**Comment by bradvogel** [comment]
@randallm working on this

**Comment by manast** [comment]
It would be great if we share his thoughts before starting the implementation.

**Comment by manast** [comment]
I think that the easiest thing to test is to change the listener from line https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L839 to line https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L865
And use a special listener that only dispatches an incoming event to one of its listeners. Since we only listen when the worker actually is idling, and only dispatches to one of the workers, the problem should be solved. My only reservation is that maybe some event could be missed, but maybe it is just paranoia.  After we return from https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L855 I think that no event would have time to be received before we setup the listener since we are in the same event loop.

**Comment by randallm** [comment]
On a high level, my thoughts were basically what @bradvogel mentioned in the OP of having the queue listen for added job events rather than keeping the existing job dispatching system and developing a special listener. Obviously this PR is proof of concept only, but I'm curious as to your thoughts on the feasibility of this approach.

**Comment by bradvogel** [comment]
@manast does https://github.com/OptimalBits/bull/pull/559/files make sense at a high level? So there's only one function that calls the moveToActive script and it only happens on bootup and when a global 'added' event is received. Let us know what you think about this approach and we'll implement it.

**Comment by randallm** [comment]
We observed a return to Bull 1.x Node CPU process utilization on average stress testing this patch. We used aws t2.micro vps instances, comparing 3.0.0rc2 to this PR with concurrency level 1000 and many jobs coming in at once:

Version: bull@3.0.0rc2
 
Producer.js: t2.micro (i * 100 delay)
Consumer t2.medium, running cluster (2 cores), concurrency 1000
Redis CPU 12%
Consumer CPU 50%
 
Version: bull@1.0.0 (mixmax fork)
 
Producer.js: t2.micro (i * 100 delay)
Consumer t2.medium, running cluster (2 cores), concurrency 1000
Redis CPU 1%
Consumer CPU 5%
 
Version: bull @mixmaxhq/bull:randall
 
Producer.js: t2.micro (i * 100 delay)
Consumer t2.medium, running cluster (2 cores), concurrency 1000
Redis CPU 1%
Consumer CPU 7%

**Comment by manast** [comment]
Sorry for having being a bit silent. I got some stomach issue that has knocked me out. 
Yeah, it looks like a good approach. Just after a brief review (and risking overlooking something), the major drawback with this approach is that in the case of having a queue with many jobs waiting, instead of directly processing the next job it will need to wait for the "added" event, which may introduce some latency. No idea if this latency is negligible in most cases, it would need to be tested.

**Comment by manast** [comment]
Btw, I will push later today what I did for the lazy event listeners as well, even if not all tests are passing yet you could take a look at it.

**Comment by randallm** [comment]
> Sorry for having being a bit silent. I got some stomach issue that has knocked me out.

My condolences 😞, and no worries at all!

> instead of directly processing the next job it will need to wait for the "added" event, which may introduce some latency

My understanding of this patch is that after `_this.processJob` resolves, we call `_this.maybePromiseJob` again. Because of this, for the next job to be processed we do not actually need to wait for the "added" event for the next job to begin processing.

> I will push later today what I did for the lazy event listeners as well, even if not all tests are passing yet you could take a look at it.

Absolutely :)

**Comment by manast** [comment]
https://github.com/OptimalBits/bull/pull/565

**Comment by manast** [comment]
@randallm ok, if there is no extra delay it seems as this approach does not have any drawbacks that I can see.
